### ISSUE TITLE
GDAL fails to open jp2 when compiled with Visual Studio 2019 due to Windows header bug

### ISSUE THREAD

**Issue by kilodyne** [OSGeo/gdal#3351]
This isn't directly a GDAL problem but I think it's worth documenting as it's very annoying and I wasted a bit of time tracking it down.

We built GDAL 3.2.1 and Jasper 1.900.1.uuid with Visual Studio 2019. We've previously built with VS2008 and VS2013 without issue. With VS2019, we had an issue with not being able to create jp2 files. In JPEG2000Dataset::Open, `poDS->nBands` would be corrupted at this line: `poDS->nBands = static_cast<int>(box->data.ihdr.numcmpts);`

`numcmpts` is defined as `uint_fast16_t`.

stdint.h in Windows has this definition: `typedef unsigned int uint_fast16_t;`

The GDAL problem was fixed by changing that definition to: `typedef unsigned short uint_fast16_t;`

I reported this to Microsoft although they closed a previous report about it 2 years ago.

**Comment by rouault** [comment]
I don't understand why you get an issue. If you build GDAL and Jasper with the same compiler, uint_fast16_t should be the same data type. Be it uint16 or uint32 shouldn't matter.

**Comment by kilodyne** [comment]
You're right. I had initially thought Jasper was doing something that relied on it being 16 bits but that's not the case.

It turns out the problem is that Jasper has it's own correct definition of uint_fast16_t in jas_types.h. My Jasper build and previous builds I did of GDAL were using that. With newer GDAL, at least since 3.1.2, it started picking up Windows stdint.h before jas_types.h resulting in a 32 bit value in GDAL but 16 bit in Jasper.

My Jasper build is being done with a migrated Visual Studio solution originally based on the workspace included in the 1.900.1.uuid archive. It didn't think Windows had stdint.h. I rebuilt Jasper with HAVE_STDINT_H defined and reverted stdint.h to have the 32 bit definition for uint_fast16_t and that fixes the issue as well. Upstream Jasper is now using cmake so this shouldn't be an issue there.
